Enable `copy` to be used with MatrixRef

albestro opened the pull request:
Basic implementation (mainly copy-paste) for a copy method involving `MatrixRef`s.

Currently `Matrix` and `MatrixRef` are still two independent objects, and this implies that we would have to duplicate all the code for the `Matrix` (more or less, e.g. `MatrixRef` might have an offset which creates some new corner cases) the same for `MatrixRef`.

This represents a problem
- from the code point of view: with duplicated code, exponential number of combinations (if A,B,C are 3 matrix-like objects there are 8 combinations of `Matrix` vs `MatrixRef` calls), trivial template solution might costs a lot in terms of compile times and binaries sizes.
- but also for tests: should I take almost all `Matrix` tests and replicate them for `MatrixRef`?!

Nothing new, but since it does not sound useful to spend time replicating things like that, in this PR I started with just the basic implementation of the `copy` and as soon as we take a decision on how to proceed, I will try to apply here too. 😉 

Waiting for your ideas and suggestions!

EDIT:
After a discussion with others, we opted for using `copy` as a "testbed" also for the sub-matrix/pipeline mechanisms, as a "real" use-case. This means that test for `copy` will be more "extensive" wrt other algorithms (e.g. gemm) and we will have:
- tests for Matrix: just full matrix
- tests for MatrixRef and Pipeline (i.e. Matrix)
  - sub-matrix
  - full-matrix

For the rest of the algorithms (so `gemm` included), we will just test the with MatrixRef both full and sub-matrices cases (where applicable).

msimberg commented:
Just for the record, I will benchmark these changes: https://github.com/eth-cscs/DLA-Future/compare/master...msimberg:DLA-Future:matrix-ref-algorithm-input-shortcut-trivial-ref. Hopefully by early next week we'll know more about if we can possibly only instantiate algorithms with `MatrixRef` and then you could update this PR to do the same.

albestro commented on test/unit/matrix/test_copy.cpp:
```diff
@@ -0,0 +1,360 @@
+
+TYPED_TEST_SUITE(MatrixCopyTest, MatrixElementTypes);
+
+// TODO local test
```
note-to-self: cover also non-distributed test-cases

msimberg commented on include/dlaf/matrix/copy.h:
```diff
@@ -77,6 +78,32 @@ void copy(Matrix<const T, Source>& source, Matrix<T, Destination>& dest) {
+  namespace ex = pika::execution::experimental;
+
+  if constexpr (Source == Destination) {
+    if (&src == &dst)
```
I don't know if this needs to be resolved in this PR, but since this compares `MatrixRef` addresses and not `Matrix` addresses this could in the worst case lead to deadlocks if the `MatrixRef`s refer to the same `Matrix` (`read` and `readwrite` into the same task). I think one step in the right direction would be to have `operator==` on `MatrixRef` (which compares the addresses of the contained `Matrix` and sub distributions), but I'm not sure if it covers all possible issues.

I'd be fine with dealing with this outside of this PR.

albestro replied:
That's a good point. I did a no-brainer copy-paste, but actually it hides a dangerous thing. How should we address this? Immediately or just open an issue? @rasolca @msimberg

rasolca replied:
I think that Submatrix pipeline should be considered as well, so it is better to work on it separately.
However it should be taken care as the tridiagonal solver might be impacted by it.
(device <-> host mirror copies in the case of multicore are skipped as they are the same matrix, however after the cleanup they will be two equal ref but different objects)

msimberg replied:
https://github.com/eth-cscs/DLA-Future/issues/1035

albestro replied:
Thanks @msimberg for opening the issue!

I'll not resolve this comment hoping it will be easier in the future to spot it in case of need of reference.

rasolca commented on test/unit/matrix/test_copy.cpp:
```diff
@@ -0,0 +1,384 @@
+
+bool isInSub(const GlobalElementIndex& ij, const GlobalElementIndex& origin,
+             const GlobalElementSize& sub_size) noexcept {
+  return ij.row() >= origin.row() && ij.col() >= origin.col() &&
+         ij.isIn(sizeFromOrigin(origin + sub_size));
```
Might be helpful to add this helper to indices directly?

albestro replied:
see 38f96c9

rasolca reviewed (commented):
Please start to use `snake_case` for functions.

albestro commented:
cscs-ci run

albestro commented on test/unit/matrix/test_copy.cpp:
```diff
@@ -0,0 +1,365 @@
+  return sub_origin == GlobalElementIndex{0, 0} && sub_size == dist_full.size();
+}
+
+template <class ElementGetter>
+auto subValues(ElementGetter&& fullValues, const GlobalElementIndex& offset) {
+  return [fullValues, offset = sizeFromOrigin(offset)](const GlobalElementIndex& ij) {
+    return fullValues(ij + offset);
+  };
+}
+
+template <class OutsideElementGetter, class InsideElementGetter>
+auto mixValues(const GlobalElementIndex& offset, const GlobalElementSize& sub_size,
+               InsideElementGetter&& insideValues, OutsideElementGetter&& outsideValues) {
+  return [outsideValues, insideValues, offset, sub_size](const GlobalElementIndex& ij) {
+    if (ij.isInSub(offset, sub_size))
+      return insideValues(ij - common::sizeFromOrigin(offset));
+    else
+      return outsideValues(ij);
+  };
+}
```
These helpers might be useful elsewhere, and I will probably move them in some "common" place so that they can be used in other tests too. (Probably this will happen in #969)

albestro commented:
cscs-ci run

rebased on new master (for getting matrix base snake_case) and addressed snake_case in implementation too.

after this it is ready to be merged 👍🏻
